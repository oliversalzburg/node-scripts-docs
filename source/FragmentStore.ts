import fs from "fs/promises";
import { Tokens, marked } from "marked";
import path from "path";
import { DOCUMENTATION_PENDING_DEFAULT } from "./FragmentRenderer";

export const DOCS_FRAGMENTS_DEFAULT_LOCATION = "docs/reference/Repository Scripts/";

export class DocumentationFragment {
  descriptionMarkdown: string;
  filename: string;

  constructor(filename: string, descriptionMarkdown = DOCUMENTATION_PENDING_DEFAULT) {
    this.filename = filename;
    this.descriptionMarkdown = descriptionMarkdown;
  }
}

export class FragmentStore {
  private fragmentDirectory: string;

  public fragments = new Map<string, DocumentationFragment>();

  constructor(fragmentDirectory: string) {
    this.fragmentDirectory = fragmentDirectory;
  }

  async loadFragment(fragmentFilename: string) {
    const assumedScriptName = FragmentStore.fragmentFilenameToScript(fragmentFilename);
    if (!assumedScriptName) {
      throw new Error(`Unable to interpret fragment file name '${fragmentFilename}'!`);
    }

    const existingDocumentation = await fs.readFile(
      path.resolve(this.fragmentDirectory, fragmentFilename),
      "utf-8",
    );
    const parsed = marked.lexer(existingDocumentation);

    // Expect parsed to be an array of [heading, list]
    const items = (parsed?.[1] as Tokens.List)?.items;
    // Expect items to be [list_item, list_item, list_item]
    const descriptionItem =
      items &&
      items.find(
        item => 1 <= item.tokens.length && (item.tokens[0] as Tokens.Text).text === "Description:",
      );
    if (!descriptionItem) {
      throw new Error(
        `Unable to find description item in documentation fragment at '${fragmentFilename}'!`,
      );
    }

    const descriptionRaw = descriptionItem.tokens
      .slice(2)
      .map(token => token.raw)
      .join("");

    this.fragments.set(
      assumedScriptName,
      new DocumentationFragment(fragmentFilename, descriptionRaw),
    );
  }

  getExistingDescription(scriptName: string) {
    if (!this.fragments.has(scriptName)) {
      return undefined;
    }

    return this.fragments.get(scriptName)!.descriptionMarkdown;
  }

  static scriptToFragmentFilename(scriptName: string) {
    // $ needs to be duplicated, as it's otherwise interpreted as part of $1,$2,$3,... references
    return `.${scriptName.replace(/\:/g, "$$$$")}.md`;
  }

  static fragmentFilenameToScript(fragmentName: string) {
    const matches = /\.(.+?)\.md/.exec(fragmentName);
    if (!matches) {
      return null;
    }

    return matches[1].replace(/\$\$/g, ":");
  }

  static async exists(storePath: string) {
    try {
      await fs.stat(storePath);
    } catch {
      return false;
    }

    return true;
  }
}
