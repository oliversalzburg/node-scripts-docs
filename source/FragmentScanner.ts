import { globby } from "globby";
import { FragmentStore } from "./FragmentStore.js";

/**
 * Loads any fragments from the given directory and places them into a fragment store.
 * @param fragmentDirectory - The directory from which to load fragments.
 * @returns A fragment store.
 */
export const loadFragments = async (fragmentDirectory: string) => {
  const fragmentStore = new FragmentStore(fragmentDirectory);
  const fragments = await globby(FragmentStore.scriptToFragmentFilename("*"), {
    cwd: fragmentDirectory,
  });
  for (const fragment of fragments) {
    await fragmentStore.loadFragment(fragment);
  }

  return fragmentStore;
};
