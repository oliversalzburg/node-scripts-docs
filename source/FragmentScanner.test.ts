import path from "node:path";
import { loadFragments } from "./FragmentScanner.js";
import { DOCS_FRAGMENTS_DEFAULT_LOCATION } from "./FragmentStore.js";

it("loads existing fragments as expected", async () => {
  const fragmentStore = await loadFragments(
    path.join("test/fixtures/default", DOCS_FRAGMENTS_DEFAULT_LOCATION),
  );
  expect(fragmentStore.fragments).toMatchSnapshot();
});
