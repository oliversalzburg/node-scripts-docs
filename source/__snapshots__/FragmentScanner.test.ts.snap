// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`loads existing fragments as expected 1`] = `
Map {
  "build" => DocumentationFragment {
    "descriptionMarkdown": "Build the TypeScript sources.",
    "filename": ".build.md",
  },
  "docs:build" => DocumentationFragment {
    "descriptionMarkdown": "Build the documentation website locally.",
    "filename": ".docs$$build.md",
  },
  "docs:scripts" => DocumentationFragment {
    "descriptionMarkdown": "Rebuilds the \`node-scripts-docs\` documentation for this project itself.",
    "filename": ".docs$$scripts.md",
  },
  "docs:serve" => DocumentationFragment {
    "descriptionMarkdown": "Runs the mkdocs-material development server.",
    "filename": ".docs$$serve.md",
  },
  "npm:publish:patch" => DocumentationFragment {
    "descriptionMarkdown": "_documentation pending_",
    "filename": ".npm$$publish$$patch.md",
  },
  "npm:publish" => DocumentationFragment {
    "descriptionMarkdown": "Build the latest sources and then \`npm publish\` the current state.",
    "filename": ".npm$$publish.md",
  },
  "nsd-debug" => DocumentationFragment {
    "descriptionMarkdown": "Build the latest sources and then use the build output to execute your command. Parameters are passed through by \`yarn\`.",
    "filename": ".nsd-debug.md",
  },
  "test:coverage" => DocumentationFragment {
    "descriptionMarkdown": "_documentation pending_",
    "filename": ".test$$coverage.md",
  },
}
`;
